File: src/data/courses/dsa/graphs/mst-advanced/alien-dictionary/code/cpp/alien_dictionary.cpp
#include <vector>
#include <string>
#include <queue>
#include <unordered_set>
#include <unordered_map>
using namespace std;
// ============================================================
// Solution 1: DFS topological sort on character graph — O(C)
// ============================================================
class Solution1 {
    unordered_map<char,unordered_set<char>>adj;
    unordered_map<char,int>color;string result;bool hasCycle=false;
    void dfs(char u){color[u]=1;for(char v:adj[u]){if(color[v]==1){hasCycle=true;return;}if(color[v]==0)dfs(v);if(hasCycle)return;}
        color[u]=2;result+=u;}
public:
    string alienOrder(vector<string>&words){
        for(auto&w:words)for(char c:w){adj[c];color[c]=0;}
        for(int i=0;i<(int)words.size()-1;i++){
            string&a=words[i],&b=words[i+1];
            if(a.size()>b.size()&&a.substr(0,b.size())==b)return"";
            for(int j=0;j<min(a.size(),b.size());j++)if(a[j]!=b[j]){adj[a[j]].insert(b[j]);break;}
        }
        for(auto&[c,_]:adj)if(color[c]==0)dfs(c);
        if(hasCycle)return"";
        reverse(result.begin(),result.end());return result;
    }
};

// ============================================================
// Solution 2: Kahn's BFS — O(C) where C = total chars
// ============================================================
class Solution2 {
public:
    string alienOrder(vector<string>&words){
        unordered_map<char,int>inDeg;unordered_map<char,unordered_set<char>>adj;
        for(auto&w:words)for(char c:w){inDeg[c];adj[c];}
        for(int i=0;i<(int)words.size()-1;i++){
            string&a=words[i],&b=words[i+1];
            if(a.size()>b.size()&&a.substr(0,b.size())==b)return"";
            for(int j=0;j<min(a.size(),b.size());j++)if(a[j]!=b[j]){if(!adj[a[j]].count(b[j])){adj[a[j]].insert(b[j]);inDeg[b[j]]++;}break;}
        }
        queue<char>q;for(auto&[c,d]:inDeg)if(d==0)q.push(c);
        string result;while(!q.empty()){char c=q.front();q.pop();result+=c;for(char v:adj[c])if(--inDeg[v]==0)q.push(v);}
        return result.size()==inDeg.size()?result:"";
    }
};

File: src/data/courses/dsa/graphs/mst-advanced/alien-dictionary/code/python/alien_dictionary.py
# ============================================================
# Alien Dictionary
# ============================================================
from collections import defaultdict, deque
class Solution:
    def alienOrder(self, words):
        adj = defaultdict(set)
        in_deg = {c: 0 for w in words for c in w}
        # Compare adjacent words
        for i in range(len(words) - 1):
            w1, w2 = words[i], words[i+1]
            if len(w1) > len(w2) and w1[:len(w2)] == w2: return ""
            for c1, c2 in zip(w1, w2):
                if c1 != c2:
                    if c2 not in adj[c1]:
                        adj[c1].add(c2); in_deg[c2] += 1
                    break
        # Kahn's
        queue = deque(c for c in in_deg if in_deg[c] == 0)
        result = []
        while queue:
            c = queue.popleft(); result.append(c)
            for nb in adj[c]:
                in_deg[nb] -= 1
                if in_deg[nb] == 0: queue.append(nb)
        return ''.join(result) if len(result) == len(in_deg) else ''
